File: Assets/Scripts/Domain/Runtime/Views/Path.cs
using UnmappedIsland.Domain.Defs;

namespace UnmappedIsland.Domain.Runtime.Views
{
    /// <summary>
    /// 道（locations.yamlのpath object_def）に対する、UI/ゲームロジック向けの型付きビュー。
    /// World/PlayerCharacter/Location と同じ理由で継承ではなくラップにしている。
    ///
    /// travel_minutes/required_progress/destination_idはいずれも生成時（IslandSpawner）にインスタンス
    /// ごとへ上書きされる値であり、このビューは読み取りとtravelアクションの実行だけを提供する。
    /// 移動そのもの（actorの所属差し替え・時間消費）はYAML側のtravelアクション（move+duration）が担う。
    /// </summary>
    public sealed class Path
    {
        public WorldObject Instance { get; }

        private readonly int travelMinutesId = -1;
        private readonly int requiredProgressId = -1;
        private readonly int destinationIdId = -1;

        public Path(WorldObject instance, NameRegistry propertyNames)
        {
            Instance = instance;
            travelMinutesId = IdOrMissing(propertyNames, "travel_minutes");
            requiredProgressId = IdOrMissing(propertyNames, "required_progress");
            destinationIdId = IdOrMissing(propertyNames, "destination_id");
        }

        /// <summary>未登録の名前は-1（LocalIndexMap.Missing扱い）にする（TryGetIdのout値は失敗時0＝
        /// 別の名前の有効なIDになってしまうため、そのままでは使えない）。</summary>
        private static int IdOrMissing(NameRegistry names, string name) =>
            names.TryGetId(name, out int id) ? id : -1;

        /// <summary>移動時間（分）。</summary>
        public int TravelMinutes => Instance.GetEffectiveValue(travelMinutesId);

        /// <summary>発見に必要な、親の土地の探索進捗。</summary>
        public int RequiredProgress => Instance.GetEffectiveValue(requiredProgressId);

        /// <summary>移動先LocationのインスタンスID。</summary>
        public int DestinationInstanceId => Instance.GetEffectiveValue(destinationIdId);

        /// <summary>この道を通って移動する（YAML側のtravelアクション: 未発見なら不成立、成功なら
        /// actorが移動先へ移り、travel_minutes分の時間が進む）。</summary>
        public bool Travel(WorldObject actor, WorldSession session) =>
            Instance.TryExecuteAction("travel", actor, session);
    }
}

File: src/domain/views/Path.ts
import type { WorldCodex } from '../WorldCodex';
import type { WorldRuleVocabulary } from '../WorldVocabulary';
import type { WorldObject } from '../WorldObject';

/**
 * 道（locations.yamlのpath object_def）に対する、UI/ゲームロジック向けの型付きビュー。Worldと同じ理由で継承では
 * なくラップにしている。
 *
 * インスタンスごとの値（行き先・所要時間・要る進捗・戻る道）は生成時にIslandSpawnerが書き込む。
 * 移動そのもの（actorの所属差し替え・時間消費）はYAML側のtravelアクションが担う。
 */
export class Path {
  readonly instance: WorldObject;

  private readonly words: WorldRuleVocabulary;

  constructor(instance: WorldObject, codex: WorldCodex) {
    this.instance = instance;
    this.words = codex.vocabulary.world;
  }

  /** 移動時間（分）。 */
  get travelMinutes(): number {
    return this.instance.tryGetProperty(this.words.travelMinutesId)?.getEffectiveValue() ?? 0;
  }

  /** 発見に必要な、親の土地の探索進捗。 */
  get requiredProgress(): number {
    return this.instance.tryGetProperty(this.words.requiredProgressId)?.getEffectiveValue() ?? 0;
  }

  /** 移動先LocationのインスタンスID。 */
  get destinationInstanceId(): number {
    return this.instance.tryGetProperty(this.words.destinationIdId)?.getEffectiveValue() ?? 0;
  }

  /**
   * 移動先のLocation。世界のツリーから引く（MoveEffectの移動先の解決と同じ引き方）ので、
   * 呼び出し側はインスタンスIDから実体を辿る手順を知らなくてよい。まだ実体化していなければundefined。
   */
  get destination(): WorldObject | undefined {
    return this.instance.findRoot().findDescendantByInstanceId(this.destinationInstanceId);
  }

  /** 移動先の土地にある、こちらへ戻る道のインスタンスID（辺の両端の道は互いを指す）。 */
  get returnPathInstanceId(): number {
    return this.instance.tryGetProperty(this.words.returnPathIdId)?.getEffectiveValue() ?? 0;
  }

  /** この道を通って移動する（YAML側のtravelアクション: 未発見なら不成立、成功ならactorが移動先へ移り、travel_minutes分の時間が進む）。 */
  travel(actor: WorldObject | undefined): boolean {
    return this.instance.tryGetAction(this.words.travelAction, actor)?.tryExecute() === true;
  }
}
